### PR TITLE
Introduce SAPA Guide Map

### PR DESCRIPTION
**Pull request by abhishek1508**
### Description

Fixes https://github.com/mapbox/navigation-sdks/issues/1412
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/6236
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/6238

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt:201]:
```suggestion
internal fun NavigationStatus.getCurrentBannerInstructions(
```

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt:30]:
```suggestion
class MapboxRestAreaApi @JvmOverloads constructor(
```

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt:36]:
```suggestion
        private const val ACCESS_TOKEN_QUERY = "?access_token="
```

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt:58]:
```suggestion
                    errorMessage = "No service/parking area guide map available for current maneuver.",
```
I don't think we should freely introduce a "sapa" into our docs and dictionary. Could we be more explicit across the docs?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/SapaProcessor.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/SapaProcessor.kt:155]: This exception should passed down to `SapaMapError`.

**Review by LukasPaczos — COMMENTED** [review]
The whole pipeline for `MapboxSignboardApi`, `MapboxJunvtionApi` and now the new proposed `MapboxSapaApi` is nearly identical. The only significant difference is that each of them checks `bannerComponent.subType()` to be either `JCT`, `SIGNBOARD`, or now `SAPAGUIDEMAP`. An addition of the `SAPAGUIDEMAP` effectively costs us more than a 1000 lines of code to write and maintain, and since we already know that there's more guidance view sub-types in the pipeline, I think we should discuss how do we expand this feature going forward.

Let's consider timelines, I don't necessarily want to block this work if it's time sensitive, since we've already done it. However, I'd like us to strongly consider deprecating the above-mentioned classes, as well as the one introduced here, in favor of a `MapboxImageGuidanceApi` and `MapboxImageGuidanceView` (or something along these lines) that would expose distinct methods to download any of the sub-types and reuse the rest of the setup.

File: examples/src/main/java/com/mapbox/navigation/examples/core/MapboxJunctionActivity.kt
Comment (abhishek1508) [reply examples/src/main/java/com/mapbox/navigation/examples/core/MapboxJunctionActivity.kt:174]: Will be reverted before merging.

File: examples/src/main/java/com/mapbox/navigation/examples/core/MapboxJunctionActivity.kt
Comment (abhishek1508) [reply examples/src/main/java/com/mapbox/navigation/examples/core/MapboxJunctionActivity.kt:163]: Will be reverted before merging.

File: examples/src/main/java/com/mapbox/navigation/examples/core/MapboxJunctionActivity.kt
Comment (abhishek1508) [reply examples/src/main/java/com/mapbox/navigation/examples/core/MapboxJunctionActivity.kt:334]: Will be reverted before merging.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/sapaguidemap/api/MapboxSapaApi.kt:58]: Removed all usage of SAPA from this PR. Let me know.
@LukasPaczos

**Comment by abhishek1508** [comment]
This PR is blocked on https://github.com/mapbox/mapbox-navigation-native/releases/tag/114.0.0. NN `v114` introduces https://github.com/mapbox/mapbox-navigation-native/pull/6186 that is required. 

cc @LukasPaczos

**Comment by abhishek1508** [comment]
https://github.com/mapbox/mapbox-navigation-android/pull/6276 landed and now the PR is complete without any dependencies. Let me know if any other changes are required.

cc @LukasPaczos

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt:36]: Can we be sure that map URLs will never have any query parameters? If they did, appending `?` would break the URL.

Let's make this more robust and use `?` if there are no query params and `&` if there already are some.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/model/GuideMapError.kt
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/model/GuideMapError.kt:10]: Remove `Exception` or `Throwable`? It was `String` and `Throwable` earlier, but changed it to `exception` based on https://github.com/mapbox/mapbox-navigation-android/pull/6239#discussion_r959503936

@LukasPaczos

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt:36]: > Can we be sure that map URLs will never have any query parameters

You mean URLs will never end with `?` or `&`?

> Let's make this more robust and use ? if there are no query params and & if there already are some.
This would then be applicable to `Signboards` and `Junctions` as well. I am not fully convinced that this check is required, since we Nav API is sure that URLs will look like they do today and are not gonna change. 

@LukasPaczos

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/model/GuideMapError.kt
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/model/GuideMapError.kt:10]: Remove the `Exception`. You can pass `Exception` as a `Throwable`.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt:36]: I think we should employ a little bit of defensive programming and assume that we don't know how the base URL looks like when we're appending a new query parameter.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt:222]: This should never happen. Although added log statements but still would have to return `null`

**Comment by abhishek1508** [comment]
@LukasPaczos Changes are addressed. Please take another look when you have a chance.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:17]: `?` already marks the first query parameter, replacing it would break the URL.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:26]: Replacing `?` with `&` would break the URL, there can only be on `?` to start the query collection.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:28]: Just because the URL doesn't end with `?` or `&` doesn't mean there are any query parameters already.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:19]: This is incorrect. We are not verifying whether there are any existing query parameters, and rarely would a URL end with `?` or `&`.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:13]: Why a different case for `JCT`? Logic for appending a query parameter should be universal for all and any URLs.

**Comment by LukasPaczos** [comment]
Here's an example of how to work with a URL: https://github.com/mapbox/mapbox-navigation-android/blob/e3e78f48abb593b296f544ea93af49e594866011/libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt#L22-L29

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:17]: For junction views, Nav ApI network request demands `&access_token=value` whereas for signboards and other subTypes it demands `?access_token=value`

I have raised this earlier with the directions team to keep it consistent.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:13]: https://github.com/mapbox/mapbox-navigation-android/pull/6239#discussion_r964746867

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/GuidanceViewUrlMatcher.kt:17]: Could you provide example URLs? This is very strange.

**Comment by abhishek1508** [comment]
The thing that I am trying to do with the URL here is making sure that it doesn't end with `?` or `&`. There ain't any problem with the URL if it already contains a collection of query. I am trying to address the problem if the guidance view URL is returned as "https://abc.mapbox.com/?". Based on the current logic, the API would end up adding `?access_token=value` to the URL making it "https://abc.mapbox.com/??access_token=value"

At least that's what I understood from your earlier comment about being defensive. However if the intention is to add query parameters using available APIs then that's different. 

@LukasPaczos

**Comment by LukasPaczos** [comment]
A URL ending with a `?` is not going to be a problem for the backend. As far as I see the existing code, for example, would mutate a `https://abc.mapbox.com/?` into `https://abc.mapbox.com/&accesss_token=token` which is incorrect.

**Comment by LukasPaczos** [comment]
All we need to do is produce a valid response by adding a query parameter, regardless of what was in the URL beforehand (whether it had any queries or already ended with `?`).

**Comment by abhishek1508** [comment]
> A URL ending with a `?` shouldn't be a problem for the backend. As far as I see the existing code, for example, would mutate a `https://abc.mapbox.com/?` into `https://abc.mapbox.com/&accesss_token=token` which is incorrect.

This mutation is correct because `Junction View` accepts `https://abc.mapbox.com/&accesss_token=token`. Whereas `subType` with any other types other than JunctionView accepts `?access_token=token`.

**Comment by abhishek1508** [comment]
> 

> Here's an example of how to work with a URL:
> 
> https://github.com/mapbox/mapbox-navigation-android/blob/e3e78f48abb593b296f544ea93af49e594866011/libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt#L22-L29

This is addressed. Please take another look when you get a chance

@LukasPaczos

File: libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/SvgUtil.kt
Comment (LukasPaczos) [inline libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/SvgUtil.kt:68]: This fully duplicates the below function. Could we parametrize `SVGExternalFileResolver` instead (make it nullable)?

**Review by LukasPaczos — APPROVED** [review]
Code wise LGTM, I trust you had a chance to test against the production endpoint as well to verify E2E.

**Comment by abhishek1508** [comment]
> Code wise LGTM, I trust you had a chance to test against the production endpoint as well to verify E2E.

Yes I tested using the following 2 pair of coordinates:

1. `137.76136788022933,34.83891088143494;137.75220947550804,34.840924660770725`
2. `137.7595462864562,34.83932704694948;137.75777196578724,34.83989897684816`

File: libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/SvgUtil.kt
Comment (abhishek1508) [reply libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/SvgUtil.kt]: Fixed.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/api/MapboxRestAreaApi.kt:121]: Is this supposed to be called from the main thread? I don't see any proxying to the worker thread on our side, is it done on the NN side?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/model/RestAreaGuideMapError.kt
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/model/RestAreaGuideMapError.kt:8]: Should we add `toString`, `hashCode` and `equals`? Not sure about the last two but I think that `toString` would be useful.

**Pull request merged**
